File: js/hash/hash.js
console.log('Доступ к данным')
console.log('a = {a:1}: ', a = {a:1})
console.log('a.a: ', a.a)
console.log('a["a"]: ', a["a"])
console.log('a.z: ', a.z)
console.log('a["z"]: ', a["z"])

console.log('Слияние')
console.log('a = {a:1}: ', a = {a:1})
console.log('b = {b:2}: ', b = {b:2})
console.log('c = {}: ', c = {})
console.log('Object.assign({},a,b): ', Object.assign({},a,b))
console.log('c={a:"перез",b:"аписал"}: ', c={a:"перез",b:"аписал"})
console.log('Object.assign({},a,b,c): ', Object.assign({},a,b,c))

console.log('Клонирование (то же что слияние с {})')
console.log('a = {a:1}: ', a = {a:1})
console.log('b = Object.assign({},a): ', b = Object.assign({},a))
console.log('a.a=2: ', a.a=2)
console.log('a.a: ', a.a)
console.log('b.a: ', b.a)

console.log('Перебор')
console.log('for (var key in b) {...')
for (var key in b) {
  console.log('key: ', key)
  console.log('b[key]: ', b[key])
}

console.log('Удаление свойства')
console.log('a = {a:1}: ', a = {a:1})
console.log('delete a.a: ', delete a.a)
console.log('a: ', a)

console.log('Ключи хэш')
console.log('Object.keys({a:1,b:null,c:undefined}): ', Object.keys({a:1,b:null,c:undefined}))

// Длинна хэш
console.log('Длинна хэш')
console.log('Object.keys({a:1,b:null,c:undefined}).length: ', Object.keys({a:1,b:null,c:undefined}).length)


// var a = {
//   'id': '1',
//   'name': 'asd'
// }
// a.map(function(k,v){
//   console.log('k: ', k)
// })

File: ruby/hash/hash.rb
# encoding: UTF-8
require_relative '../colorize/colorize'

arr = 'asd'
h = {
  :r => "Камень",
  :arr => arr,
  :Mike => 214134,
  :jes => '1432352',
  0 => ['asd'],
  asdf: ['asd'],
  'asdf' => ['asd'],
}
p h

# Добавить новый элемент
h[3] = '3'
# h[:mike] = 'test'
p h
p h[:Mike]

# Имеется ли элемент с ключем
if h.key? :mike
  p h[:mike]
end
p h[0]
p h[3]

# Перебор
p 'Перебор'
h.each do |key,value|
  print key.to_s + ' - '
  print value.to_s
  print "\n"
end

# Количество элементов
p h.count

# Удалить элемент с ключем :a
h.delete("a")

p nil&&nil

# Преобразование в hash с фильтрацией
puts 'Преобразование в hash с фильтрацией'.red
p Hash[ [1,2,3,4,5,6,7,8].select{ |node|
  node != 2
}.map{ |node|
  [node, node]
}]

puts ({a:1,b:2}).select{ |k,v| k != :a}

# Преобразование всех полей в хэше
puts 'Преобразование всех полей в хэше'.green
# Старый хэш, старые строки (преобразованные):
a = {a:'a0',b:'b0'}
a.each{ |_,v| v.gsub! '0', '1' }
print 'a: '.red; puts a
# Старый хэш, новые строки:
a = {a:'a0',b:'b0'}
a.each{ |k,v| a[k] = "#{v[0]}2" }
print 'a: '.red; puts a
# Новый хеш:
a = {a:'a0',b:'b0'}
b = Hash[a.map{|k,v| [k,v+'3'] } ]
print 'a: '.red; puts a
print 'b: '.red; puts b

# Слияние двух hash
puts 'Слияние двух hash'.red
h1 = {a:'a',b:'b'}
h2 = {a:nil,c:'cc'}
h3 = h1.merge(h2)
p h1
p h2
p h3
h1.merge! h2
p h1

# Ключи в стрингу
a = {a:'aa',c:'cc'}
# {a:'aa',c:'cc'}.stringify_keys
p Hash[a.map{|k,v|[k.to_s,v]}]
p a

p a.delete 'a'
p a.delete :a
p a

a = 'ds'
a = nil
puts "asd: #{a.to_s.red}"

